### DISCUSSION TITLE
Power switch for track power?

### DISCUSSION THREAD

**Discussion by Budman1758** [DCC-EX/CommandStation-EX#331] [Q&A] [answered]
Is it possible to add a physical switch to control main and programing tracks? Sometimes it might be handy to be able to turn on or off when not actually running locos. For instance, I am working on a camera car that's powered by the track and every time I need power I have to find a throttle to turn the track power on or off. 

This is mainly for my test and general screwing around Commandstation.  :)

Can't seem to find any info on this.

**Comment by habazut** [comment]
In an exrail script you can wait for a button press and then do POWERON or POWEROFF.

Harald.

**Comment by Asbelos** [comment]
You could always put a physical switch between the command station and the track.
Alternatively you can use a digital switch, monitor it with exrail and issue power commands as required.

**Comment by Budman1758** [comment]
>You could always put a physical switch between the command station and the track.

Thanks for those ideas. Not really the same as a switch built into the command station though. You still need "something else" to turn the track power on or off.

**Comment by FrightRisk** [comment, marked as answer]
Hi. I'm not sure you understood the suggestion to use EX-RAIL or maybe we don't understand. Have you looked at the documentation? This sounds like what a mimic panel would do. Create a myAutomation.h file and upload the CS software again. Connect a button to pin 40

//  toggle power when button is pressed.

SEQUENCE(1)
IF(40)            // push-button pressed
AT(-40)          // wait for button release
IF  ??
POWERON
SET or LATCH ??
ELSE         
POWEROFF
UNSET or UNLATCH ??
ENDIF
ENDIF

I'm on my phone so can't check this out, but maybe one of the other guys can fix my script and provide more info on what to use for a check and to toggle the state variable. You need the development release of the CS software to use the power commands.

**Reply by Budman1758** [reply]
> Hi. I'm not sure you understood the suggestion to use EX-RAIL or maybe we don't understand. Have you looked at the documentation? This sounds like what a mimic panel would do. Create a myAutomation.h file and upload the CS software again. Connect a button to pin 40

I guess I didn't understand. That sounds like it would work. Sorry for the confusion.

**Reply by SweaterMichael** [reply]
Newb here, I’m two weeks into both DCC-EX and Arduino, but this sparked my interest.  It took me two days, but with the help of [FrightRisk](https://github.com/FrightRisk)’s reply I implemented it.

First, I couldn’t find the ??, fortunately, I also wanted an LED on the case is was(am) 3D printing to indicate track power on.  I’m on a Teyleten Robot(non-branded) Mega 2560+WiFi, KOOBOOK(Deek-Robot) motor shield, both from Amazon.

I added this to CommandDistributor.cpp:
“   (main) ? digitalWrite(49, HIGH) : digitalWrite(49, LOW);”
on line 262(5.0.9 release), just after”LCD(2,F(Power %S%S…etc…”

This is actually a fancy if, then, else statement I just learned.  Apparently, it actually matters if you have an LED plugged in, because it stopped working when I unplugged mine(?).

Next(after a few hundred trys) I put this the following in myAutomation.h:
AUTOSTART
	START(99)
DONE

SEQUENCE(99)
	if(48)
		AT(-48)
		IF(49) POWERON
	ELSE POWEROFF
		ENDIF
	ENDIF
DELAY(5O)
FOLLOW(99)

I don’t understand anything about the latching in Frightrisk’s reply.  Just to be clear, I have an LED(with resistor) from GND to pin 49 and a switch from GND to 48.  I don’t understand why 48 to positive doesn’t work, maybe a DCC-EX thing(I’d love to know?)

**Comment by Asbelos** [comment]
Firstly, it's not a good idea to meddle in the code until you have a much clearer understanding of the ? : operator. Also we don't actually use the arduino digitaWrite function for performance reasons. 

Your switch must be to ground because the pin will he held positive until you ground it. If you connect the switch to positive... it doesn't change the pin input state when you switch. 

If you just want a led that lights up when power is on, you could just wire a led plus suitable resistor accross the track
 outputs.

**Reply by SweaterMichael** [reply]
That LED across the tracks bit, that's actually something that didn't occur to me, Thanks.  Do you have a better suggestion for ??(if power is on the main)? or the LATCH bit, it definitely seems to me you know what you're talking about.

**Reply by SweaterMichael** [reply]
From [https://github.com/orgs/DCC-EX/projects/7/views/9?pane=issue&itemId=18120276]

This was originally designed to latch a pushbutton without having to deal with debouncing etc... hence the name.

The clever bit ...
If you test a sensor pin that does not exist then it returns false, unless of course it is latched!
So a latch is a very convenient way of setting a flag that can be tested later..... now you have the building block to play with.

**Reply by FrightRisk** [reply]
Are you on our Discord server? That is the best place to get interactive help. On Discord, Habazut is "Haba" and Asbelos is "UKBloke". The are our lead developers ;) If you look at the commit graph here or check this web page you will realize we are one of the few organizations where you can speak directly to the creators and decision makers. Welcome aboard! https://dcc-ex.com/about/about.html#gsc.tab=0

**Reply by SweaterMichael** [reply]
Thank You! I Actually found and signed up to your Discord last night, but planned on staying quiet till I got a feel for the place and hopefully got myself out of the "knows just enough to be dangerous" phase.  My user name is the same as here.  I've only been keeping the discussion going here because I was hoping the OP would chime in with weather he was ever successful.  Thanks again and I look forward to seeing you on the Discord.

**Comment by Asbelos** [comment]
I should hope I know what I'm talking about... shouldn't be too difficult to find out why 🧚‍♂️
